File: python-packages/core/src/tsv_data_analytics/file_io_wrapper.py
"""FileReader / FileWriter class"""

import gzip
import zipfile

from omigo_core import s3_wrapper
from omigo_core import utils

class FileWriter:
    """FileWriter class to write data into files"""
    def __init__(self, output_file_name, s3_region, aws_profile):
        self.output_file_name = output_file_name
        self.data = []
        self.s3_region = s3_region
        self.aws_profile = aws_profile

    def write(self, line):
        self.data.append(line)

    def close(self):
        output_zipf = None
        output_file = None
        if (self.output_file_name.startswith("s3://")):
            content = "".join(self.data)
            bucket_name, object_key = utils.split_s3_path(self.output_file_name)
            s3_wrapper.put_s3_file_with_text_content(bucket_name, object_key, content, self.s3_region, self.aws_profile)
        else:
            # construct output file
            if (self.output_file_name.endswith(".gz")):
                output_file = gzip.open(self.output_file_name, "wt")
                # write all the content
                for line in self.data:
                    output_file.write(line) 
            elif (self.output_file_name.endswith(".zip")):
                output_zipf = zipfile.ZipFile(self.output_file_name, "w")
                output_file = output_zipf.open(self.output_file_name.split("/")[-1][0:-4], "w")
                # write all the content
                for line in self.data:
                    output_file.write(str.encode(line))
            else:
                output_file = open(self.output_file_name, "w")
                # write all the content
                for line in self.data:
                    output_file.write(line) 

            # close
            if (output_file is not None):
                output_file.close()
            if (output_zipf is not None):
                output_zipf.close()

            # set data to None
            self.data = None

class FileReader:
    """FileReader class to read data files"""
    def __init__(self, input_file_name, s3_region, aws_profile):
        pass

    def read(self):
        pass

    def close(self):
        pass
        

File: aloki/src/main/scala/omigo_core/FileIOWrapper.scala
package omigo_core
// import scala.jdk.CollectionConverters
import collection.JavaConverters._
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.util.Random
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

class FileWriter(val outputFileName: String, val s3Region: String, val awsProfile: String) {
  // TODO: Not needed probably
}

class TSVFileWriter(val s3Region: String, val awsProfile: String) {
  def save(xtsv: TSV, outputFileName: String): Unit = {
    val content = List(xtsv.getHeader(), xtsv.getData().mkString("\n")).mkString("\n")

    // check s3 or local
    if (outputFileName.startsWith("s3://")) {
      val (bucketName, objectKey) = Utils.splitS3Path(outputFileName)
      S3Wrapper.putS3FileWithTextContent(bucketName, objectKey, content, s3Region, awsProfile)
      println("file saved to: " + outputFileName)
    } else {
      // get bytes
      var barr = content.getBytes()

      // check file extensions
      if (outputFileName.endsWith(".gz")) {
        val byteStream = new ByteArrayOutputStream(barr.length)
        val zipStream = new GZIPOutputStream(byteStream)
        zipStream.write(barr)
        zipStream.close()
        barr = byteStream.toByteArray()
      } else if (outputFileName.endsWith(".zip")) {
        throw new Exception("Not supported")
      }

      // open FileOutputStream and write
      val fos = new FileOutputStream(new File(outputFileName))
      fos.write(barr)
      fos.close()
      println("file saved to: " + outputFileName)
    }
  }
}

class FileReader {
  // TODO: Not needed
}

object TSVFileWriter {
  def main(args: Array[String]): Unit = {
    val xtsv = new TSV("col1", List("value1", "value2")) 
    val writer = new TSVFileWriter(null, null)
    writer.save(xtsv, args(0))
  }
}
